packages/readmodel/src/catalogReadModelService.ts

```diff
@@ -0,0 +1,240 @@
+import { and, eq, lte } from "drizzle-orm";
+import { EService, EServiceId, WithMetadata } from "pagopa-interop-models";
+import {
+  DrizzleReturnType,
+  eserviceDescriptorAttributeInReadmodelCatalog,
+  eserviceDescriptorDocumentInReadmodelCatalog,
+  eserviceDescriptorInReadmodelCatalog,
+  eserviceDescriptorInterfaceInReadmodelCatalog,
+  eserviceDescriptorRejectionReasonInReadmodelCatalog,
+  eserviceDescriptorTemplateVersionRefInReadmodelCatalog,
+  eserviceInReadmodelCatalog,
+  eserviceRiskAnalysisAnswerInReadmodelCatalog,
+  eserviceRiskAnalysisInReadmodelCatalog,
+  eserviceTemplateRefInReadmodelCatalog,
+} from "pagopa-interop-readmodel-models";
+import { splitEserviceIntoObjectsSQL } from "./catalog/splitters.js";
+import {
+  aggregateEservice,
+  toEServiceAggregator,
+} from "./catalog/aggregators.js";
+import { checkMetadataVersion } from "./utils.js";
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function catalogReadModelServiceBuilder(db: DrizzleReturnType) {
+  return {
+    // eslint-disable-next-line sonarjs/cognitive-complexity
+    async upsertEService(
+      eservice: EService,
+      metadataVersion: number
+    ): Promise<void> {
+      await db.transaction(async (tx) => {
+        const shouldUpsert = await checkMetadataVersion(
+          tx,
+          eserviceInReadmodelCatalog,
+          metadataVersion,
+          eservice.id
+        );
+
+        if (shouldUpsert) {
```

Comment:
no change requested but for future occurrences, to reduce cognitive complexity and improve readability, you may prefer a short circuit condition like
 ```suggestion
        if (!shouldUpsert)
          return/continue;
```